sirepo/sim_db_file.py

```diff
@@ -1,48 +1,378 @@
+        self.__path.move(p)
+
+    async def _sr_post_save_from_url(self, args):
+        max_size = sirepo.job.cfg().max_message_bytes
+        size = 0
+        ok = False
+        # so update is atomic
+        t = self.__path + ".tmp"
+        try:
+            async with aiohttp.ClientSession() as s:
+                async with s.get(args.src_url) as r:
+                    if r.status != 200:
+                        if r.status == 404:
+                            return PKDict(error="not found")
+                        return PKDict(error=f"http_status={response.status}")
+                    async with aiofiles.open(t, "wb") as f:
+                        async for c in r.content.iter_chunked(_CHUNK_SIZE):
+                            size += len(c)
+                            if size > max_size:
+                                return PKDict(error=f"too large (max={max_size})")
+                            await f.write(c)
+            ok = True
+        except Exception as e:
+            pkdlog("url={} exc={} stack={}", args.src_url, e, pkdexc())
+            if isinstance(e, socket.gaierror):
+                return PKDict(error="invalid host")
+            return PKDict(error=e)
+        finally:
+            if ok:
+                t.move(self.__path)
+            else:
+                pkio.unchecked_remove(t)
+        return PKDict(size=size)
+
+    async def _sr_post_size(self, args):
+        return self.__path.size()
+
+    def __authenticate_and_path(self):
+        self.__uid = self._rs_authenticate()
+        self.__path = _uri_parse(self.request.path, uid=self.__uid)
+        if self.__path is None:
+            self.send_error(403)
+            return False
+        return self.__path
+
+    def __uri_arg_to_path(self, uri):
+        res = _uri_parse(uri, uid=self.__uid, is_arg_uri=True)
+        if res is None:
+            self.send_error(403)
+            return None
+        return res
+
+
+class SimDbUri(str):
+    def __new__(cls, sim_type, slu, basename):
+
+        if isinstance(slu, cls):
+            if basename is not None:
+                raise AssertionError(
+                    f"basename={basename} must be none when uri={slu} supplied"
+                )
+            if sim_type != slu._stype:
+                raise AssertionError(f"sim_type={sim_type} disagrees with uri={slu}")
+            return slu
+        self = super().__new__(cls, cls._uri_from_parts(sim_type, slu, basename))
+        self._stype = sim_type
+        return self
+
+    @classmethod
+    def _uri_from_parts(cls, sim_type, sid_or_lib, basename):
+        """Generate relative URI for path
+
+        Args:
+            sim_type (str): which code
+            sid_or_lib (str): simulation id. if None or `_LIB_DIR`, is library file
+            basename (str): file name with extension
+
+        Returns:
+            str: uri to be passed to SimDbClient functions
+        """
+        from sirepo import simulation_db, template
```

Comment:
which problem?

Reply:
That we have to be so careful with imports. Now there is another file (sim_db_file) where we have to remember to import in a specific way.

Reply:
sim_db_file doesn't need to be imported, which is why it is written that way. Template should not have to be a special import, actually. simulation_db is slowly getting refactored so that it doesn't have to be a special import either. Ideally, you would refer to it only via qcall, and in job agent, it would not be referred to at all (just util, const, etc.). 

So I don't think the problem is getting worse, but it is complicated to change interfaces over time as we are doing.